### ISSUE TITLE
Refactor imperative core AST

### ISSUE THREAD

**Issue by garyb** [purescript/purescript#711]

**Comment by garyb** [comment]
I know you mentioned we could just rename the JS AST @paf31, but I couldn't resist having a look to see if we could do something a bit more "core"/reusable for other targets. At the moment I have this:

``` haskell
data Expr a
  -- |
  -- A literal value
  --
  = Literal a (Literal (Expr a))
  -- |
  -- An object property or array accessor expression (prop/index, obj/array)
  --
  | Accessor a (Expr a) (Expr a)
  -- |
  -- An anonymous function value (arguments, body)
  --
  | AnonFunction a [Ident] [Statement a]
  -- |
  -- Function application
  --
  | App a (Expr a) [Expr a]
  -- |
  -- Variable reference
  --
  | Var a Ident

type Label = String

data Statement a
  -- |
  -- An expression whose value is discarded (probably causes a side effect)
  --
  = Expr (Expr a)
  -- |
  -- A function introduction (name, arguments, body)
  --
  | Function Ident [Ident] [Statement a]
  -- |
  -- A variable introduction and optional initialization
  --
  | VarIntroduction Ident (Maybe (Expr a))
  -- |
  -- A variable assignment (assignee, value)
  --
  | Assignment (Expr a) (Expr a)
  -- |
  -- While-style loop (condition, body)
  --
  | Loop (Expr a) [LoopStatement a]
  -- |
  -- A loop over an object's keys (key variable, object value, body)
  --
  | ObjectIterator Ident (Expr a) [LoopStatement a]
  -- |
  -- If-then-else statement
  --
  | IfElse (Expr a) [Statement a] (Maybe [Statement a])
  -- |
  -- Return statement
  --
  | Return (Expr a)
  -- |
  -- Throw statement
  --
  | Throw (Expr a)
  -- |
  -- InstanceOf test
  --
  | InstanceOf (Expr a) (Expr a)
  -- |
  -- Labelled statement
  --
  | Label Label (Statement a)
  -- |
  -- Commented statement
  --
  | Comment [Comment] (Statement a)

data LoopStatement a
  -- |
  -- Break statement
  --
  = Break (Maybe Label)
  -- |
  -- Continue statement
  --
  | Continue (Maybe Label)
  -- |
  -- Standard statement
  --
  | Statement (Statement a)
```

I don't know if this is the kind of route we do want to go down, but figured I'd throw it out there. It's not really expression orientated as @jdegoes suggested, but it feels like a middle ground between being too JS specific and something completely different.

The only thing in here I'm definitely unhappy with is `InstanceOf`, as we don't have a representation of classes at all in this, but if we want to desugar cases into `IfElse`s then we need something like `InstanceOf`, but perhaps it just needs to be more PureScript specific, like `IsDataConstructor Ident (Expr a)` or something?

Also would we want annotations on `Statement` as well as `Expr`?

Anyway, thoughts please.

**Comment by ikarienator** [comment]
Sorry, what is the purpose of the imperative AST? I thought it was used for optimization? Isn't this design a little bit too "high level"? I thought it would be some sort of SSA or similar, or even simply llvm.

**Comment by paf31** [comment]
I like this, but agree that the `instanceof` part is a bit unfortunate. It comes down to what to do about "tagged" values. I think we have two options:

- Tagged things don't exist in the imperative core AST. In this case, we desugar to records or something else, but perhaps pay a performance tax in JS.
- Tagged things do exist in the imperative core AST. In this case, maybe we should separate tag constructors from `Var`s as a new data constructor, to make it really clear that tagged things are special. Then we leave it up to the backend to pick a representation (prototypes, records, whatever). Ideally, we'd use a phantom type or GADT to encode this sort of info in the AST, but this makes codegen much more difficult or even impossible. 

I personally prefer the latter option.

**Comment by paf31** [comment]
Oh, also, I think annotations on statements would be nice unless there is some obstacle to adding them.

**Comment by jdegoes** [comment]
Non-extensible records can be easily compiled to high-performance struct-like constructs in many backends. Extensible records (i.e. adding fields to a record) will require a slower map-like construct. So maybe some notion of "class" or tagged "struct" / "record" could be useful.

**Comment by paf31** [comment]
Good point.

**Comment by garyb** [comment]
@ikarienator this is intended for optimization, but only for things like TCO, a little inlining, etc. We're trying to keep the JS output as human readable and somewhat close to the input where practical. That's one of the selling points of the language over using something like GHCJS actually.

The other purpose it has is to act as an intermediary step between other potential backends apart from JS, as @osa1 did some work to make a version of psc that produced Lua a while back, and there's been a little interest in a JVM target also, so having a proper non-JS-specific imperative AST will help with that.

@paf31, @jdegoes I'll address your points in a bit, but basically, :+1:. And adding annotations to statements is no problem at all.

**Comment by paf31** [comment]
I'm seriously considering making a case for removing the busted TCO implementation and promoting `purescript-tailrec` actually.

**Comment by joneshf** [comment]
> The other purpose it has is to act as an intermediary step between other potential backends apart from JS, as @osa1 did some work to make a version of psc that produced Lua a while back, and there's been a little interest in a JVM target also, so having a proper non-JS-specific imperative AST will help with that.

I think this would spur multiple backends quite a bit. I looked into writing a backend the other weekend, and It's just a bit too much work for a weekend project, for me at least.

**Comment by garyb** [comment]
@paf31: Ah yeah, I caught up with your video from the last session the other day and saw you mention that. What's wrong with it at the moment, I thought most of the kinks had been ironed out now?

We can definitely retain a notion of tagged things as per your second suggestion, we do that in `CoreFn` currently with an `IsConstructor` metadata annotation, and then there's a `Constructor` value which is a representation of a data constructor - essentially a function which produces a tagged thing. 

From the above then, I'll add a `Constructor` case to `Expr`, and `IsInstanceOf` becomes `IsTagOf Ident (Expr a)`, and we can defer the actual implementation of those things to the JS/whatever codegen.

**Comment by paf31** [comment]
I think a `--dump-imp-core` option would be nice to have too. Then people could implement backends in different languages by parsing the core representation. I created an issue for that separately.

**Comment by garyb** [comment]
Yeah, definitely :)

**Comment by garyb** [comment]
@joneshf I hope it does! There will be a bunch more issues to solve yet, like the FFI, but at least this should help.
